src/lib.rs

```diff
@@ -189,9 +160,102 @@ pub fn compile_system_policy(input_files: Vec<&str>) -> Result<String, error::Ca
+        system_names.clear();
+        for system in system_map.values() {
+            system_names.push(system.name.to_string());
+        }
+    }
+
+    // Create a dummy module to use in the dummy system if any
+    let dummy_module = &mut ValidatedModule::new(
+        CascadeString::from("module"),
+        BTreeSet::new(),
+        BTreeSet::new(),
+        None,
+        None,
+    )?;
+
+    // Add all of the type infos into the dummy module
+    // Insert the dummy module into the dummy system and insert the system into the system map
+    if let Some(mut s) = dummy_system {
+        for type_info in type_map.values() {
+            dummy_module.types.insert(type_info);
+        }
+        s.modules.insert(dummy_module);
+        system_map.insert(s.name.to_string(), s)?;
+    }
+
+    let mut system_hashmap = HashMap::new();
+    for system_name in system_names {
+        match system_map.get(&system_name) {
+            Some(system) => {
+                let mut system_type_map = compile::get_built_in_types_map()?;
+
+                // Get the type infos for the system
+                for module in &system.modules {
+                    compile::get_reduced_types(
+                        module,
+                        &mut system_type_map,
+                        &type_map,
+                        &module_map,
+                    )?;
+                }
+
+                // Generate system type aliases
+                let sys_t_aliases = compile::collect_aliases(system_type_map.iter());
+                system_type_map.set_aliases(sys_t_aliases);
+
+                let system_cil_tree = compile::get_infos(
+                    &system_type_map,
+                    &policies,
+                    &classlist,
+                    &Some(&system.configurations),
+                )?;
+
+                let system_cil = generate_cil(system_cil_tree);
 
-    errors.into_result(generate_cil(cil_tree))
+                system_hashmap.insert(system_name, system_cil);
+            }
+            None => errors.append(CascadeErrors::from(BadSystemError::new(&format!(
+                "System {} does not exist.\nThe valid systems are {}",
+                system_name,
+                system_map
+                    .values()
+                    .map(|s| s.name.as_ref())
+                    .collect::<Vec<&str>>()
+                    .join(", ")
+            )))),
+        }
+    }
+    errors.into_result(system_hashmap)
+}
+
+/// Compile all systems into a single policy
+/// The list of input files list should contain filenames of files containing policy to be
+/// compiled.
+/// The system name is the name of the output file (out.cil by default).
+/// Returns a Result containing either a string of CIL policy which is the compiled result or a
+/// list of errors.
+/// In order to convert the compiled CIL policy into a usable policy, you must use secilc.
+pub fn compile_all_system_policies(
```

Comment:
The naming here is weird.

There are 3 cases, right?

* casc provides a list of system names to build
* casc wants to build everything into one system
* casc wants to build all defined systems

This function name sounds like the third one, but I think it implements the second.

I think the API should be a separate call for each of the three above cases.  Then we have a private compile_system_policies_internal() that does all the actual work.  Each of the three public functions just get system stuff set up appropriately for the internal() function.